src/software/ai/evaluation/scoring/skills/skill_graph.cpp

```diff
@@ -0,0 +1,109 @@
+#include "software/ai/evaluation/scoring/skills/skill_graph.h"
+
+#include "software/util/generic_factory/generic_factory.h"
+
+SkillGraph::SkillGraph(std::shared_ptr<Strategy> strategy)
+{
+    auto all_skill_constructors =
+        GenericFactory<std::string, Skill,
+                       std::shared_ptr<Strategy>>::getRegisteredConstructors();
+
+    CHECK(!all_skill_constructors.empty())
+        << "No Skills registered in the Skill factory!";
+
+    std::transform(all_skill_constructors.begin(), all_skill_constructors.end(),
+                   std::back_inserter(nodes_), [&](auto skill_constructor) {
+                       return std::move(skill_constructor(strategy));
+                   });
+
+    // The size of the adjacency matrix must be one more than the size of nodes_
+    // since there is a "start" node not explicitly represented by any Skill
+    // that needs to be accounted for
+    adj_matrix_ = std::vector<std::vector<double>>(
+        nodes_.size() + 1, std::vector<double>(nodes_.size() + 1, DEFAULT_EDGE_WEIGHT));
+
+    // The first node of the sequence should always be the "start" node, which has
+    // an ID of nodes_.size()
+    sequence_ = {std::make_pair(nodes_.size(), nodes_.size())};
```

Comment:
how is this assumption enforced?

Reply:
I see so we implicitly assume that the start node is at node.size(), node.size()